### ISSUE TITLE
Collective operation timeout / SIGABRT in FullFinetuneRecipeDistributed on single-node 8-GPU run

### ISSUE THREAD

**Issue by happypanda94** [pytorch/torchtune#2836]
I am trying to check my setup by full finetuning of the Llama_3_1_8b model using the alpaca_data_cleaned.json data file. My code manages to run properly on a single GPU, so I am trying to use a server cluster node, which contains 8 GPUs. However, when running FullFinetuneRecipeDistributed on a single node with 8 GPUs, training hangs partway through and eventually aborts with a collective-operation timeout (NCCL watchdog) on one of the ranks, raising SIGABRT. 

To Reproduce:

- Checkout torchtune (latest main or v0.x).
- Prepare a JSON “instruct” dataset, e.g. alpaca_data_cleaned.json.
- Launch distributed tuning with:

```
tune run --nproc_per_node 8 full_finetune_distributed --config full_finetune_distributed .yaml

```

- Typical config excerpt:

```
batch_size: 2
device: cuda
dtype: bf16
enable_activation_checkpointing: true
epochs: 1
gradient_accumulation_steps: 1
optimizer:
  _component_: torch.optim.AdamW
  lr: 2e-5
model:
  _component_: torchtune.models.llama3_1.llama3_1_8b
dataset:
  _component_: torchtune.datasets.instruct_dataset
  data_files: /path/to/alpaca_data_cleaned.json
checkpointer:
  checkpoint_dir: /path/to/original/checkpoints/
  output_dir: /path/to/output/full_distributed

```

- Training starts, you see loss logs, then after a minute or two the progress bar freezes and you get:

> [E618 11:40:48.973] Watchdog caught collective operation timeout:
> WorkNCCL(... ALLREDUCE ...) ran for 600004 milliseconds before timing out.
> [PG ID 1 ... Rank 6] Exception detected by watchdog...
> terminate called after throwing an instance of 'c10::DistBackendError'
>   what(): ... collective operation timeout ...
> Signal 6 (SIGABRT) received by PID 48912
> 

Distributed training should proceed without NCCL timeouts. All ranks should progress through the iteration loop.

Logs:
Full excerpt of the error log

> 1|12|Loss: 1.1769…:   0%|▏| 12/3235 [01:08<5:06:16, 5.70s/it]
> [rank6]:[E618 11:40:48.9736…] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=28, OpType=ALLREDUCE, NumelIn=1, NumelOut=1, Timeout(ms)=600000) ran for 600004 milliseconds before timing out.
> [rank6]:[E618 11:40:48.997…] Exception detected by watchdog …
> [rank0]:[E618 11:40:48.115…] Watchdog caught collective operation timeout: …
> … (similar errors on ranks 0–7) …
> terminate called after throwing an instance of 'c10::DistBackendError'
>   what():  [PG ID 1 PG GUID 1 Rank 6] Process group watchdog thread terminated with exception …
> Signal 6 (SIGABRT) received by PID 48912
> 

Environment (I am only using one node in the cluster):

> 2x 20 core 2.5 GHz [Intel Xeon Gold 6248](https://ark.intel.com/content/www/us/en/ark/products/192446/intel-xeon-gold-6248-processor-27-5m-cache-2-50-ghz.html)
> 8x NVIDIA Tesla V100 GPU each with 32 GiB HBM (Driver version: 535.104.05)
> 768 GiB RAM per node
> Dual 100 Gb EDR Infiniband
> 
> PyTorch version: 2.5.1
> Built with CUDA: 12.1
> cuDNN version: 9.1.0 (90100)
> NCCL version: (2, 21, 5)
> 
> torchtune version: 0.6.1
> slurm version: 23.11.10

Additional context

- Single-node multi-GPU setup.
- Activation checkpointing enabled, offloading disabled.
- Occurs reliably after the first few iterations.
- Lowering batch size or disabling checkpointing did not help.

Any pointers on debugging NCCL timeouts or recommended config tweaks would be greatly appreciated!

**Comment by felipemello1** [comment]
hey @happypanda94, just so we can rule out any issues with the config and/or dataset, could you try running this config? https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama3_1/8B_full.yaml

Just download the model:
`
tune download meta-llama/Meta-Llama-3.1-8B-Instruct --output-dir /tmp/Meta-Llama-3.1-8B-Instruct --ignore-patterns "original/consolidated.00.pth"
`

Then run:
`
tune run --nproc_per_node 8 full_finetune_distributed --config llama3_1/8B_full
`
